<h1><center> Using Machine Learning Techniques in <br> the Detection of Fake News </center></h1>

<h2 style='font-weight:normal'><center> Spring Term Project, <b> Lisanna Lehes </b> </center></h2>

<div style="text-align: right"> 
Universidad de Huelva <br>
Facultad de CC. Experimentales <br>
Grado en Química <br>
Computational Chemistry
 </div>


#### <b>Aim of the project:</b> to create a machine learning model that would help detect fake news using Scikit-learn library and Passive Agression Algorithm.

## Part I

### 1. Importing the necessary libraries 

In [34]:
import pandas as pd
import numpy as np

### 2. Importing the dataset

In [2]:
# Importing dataset
df = pd.read_csv('/Users/Lisanna/Desktop/fake_news/train.csv')

In [3]:
df.shape

(20800, 5)

In [4]:
# Returning the first 5 rows
df.head()

id                                              title              author  \
0   0  House Dem Aide: We Didn’t Even See Comey’s Let...       Darrell Lucus   
1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...     Daniel J. Flynn   
2   2                  Why the Truth Might Get You Fired  Consortiumnews.com   
3   3  15 Civilians Killed In Single US Airstrike Hav...     Jessica Purkiss   
4   4  Iranian woman jailed for fictional unpublished...      Howard Portnoy   

                                                text  label  
0  House Dem Aide: We Didn’t Even See Comey’s Let...      1  
1  Ever get the feeling your life circles the rou...      0  
2  Why the Truth Might Get You Fired October 29, ...      1  
3  Videos 15 Civilians Killed In Single US Airstr...      1  
4  Print \nAn Iranian woman has been sentenced to...      1

### 3. Converting the '0's and '1's to 'FAKE' and 'TRUE'

‘0’ for RELIABLE article <br>
‘1’ for FAKE NEWS

In [5]:
# Accessing a group of rows and columns by label
df.loc[(df['label'] == 1) , ['label']] = 'FAKE'
df.loc[(df['label'] == 0), ['label']] = 'REAL'

In [11]:
df.head()

id                                              title              author  \
0   0  House Dem Aide: We Didn’t Even See Comey’s Let...       Darrell Lucus   
1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...     Daniel J. Flynn   
2   2                  Why the Truth Might Get You Fired  Consortiumnews.com   
3   3  15 Civilians Killed In Single US Airstrike Hav...     Jessica Purkiss   
4   4  Iranian woman jailed for fictional unpublished...      Howard Portnoy   

                                                text label  
0  House Dem Aide: We Didn’t Even See Comey’s Let...  FAKE  
1  Ever get the feeling your life circles the rou...  REAL  
2  Why the Truth Might Get You Fired October 29, ...  FAKE  
3  Videos 15 Civilians Killed In Single US Airstr...  FAKE  
4  Print \nAn Iranian woman has been sentenced to...  FAKE

### 4. Splitting the dataset to test and training data

In [6]:
labels = df.label
labels.head()

0    FAKE
1    REAL
2    FAKE
3    FAKE
4    FAKE
Name: label, dtype: object

In [72]:
labels.value_counts()

FAKE    10413
REAL    10387
Name: label, dtype: int64

Splitting the downloaded dataset into two subsets, 70% of the entries
will be used to train the model and the rest (30%) to test the model’s predictive power.

In [7]:
from sklearn.model_selection import train_test_split
# Split arrays or matrices into random train and test subsets

In [8]:
x_train, x_test, y_train, y_test = train_test_split(df['text'].values.astype('str'), labels, test_size = 0.3, random_state = 7)

In [9]:
# Random_state parameter may be provided to control the random number generator used

### 5. Using TfidfVectorizer

__TfidfVectorizer__ uses stop words from the English language. <br>
The number of times a word appears in a document is its __Term Frequency__. <br> A higher value means a term appears more often than others, and so, the document is a good match when the term is part of the search terms.


In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [16]:
tfidf_vectorizer = TfidfVectorizer(stop_words = 'english', max_df = 0.7)

__max_df__ --> ignores terms that appear in more than 70% of the documents --> used for terms that appear too frequently

### 6. Fitting and transfroming the test and training data set

In [23]:
tfidf_train = tfidf_vectorizer.fit_transform(x_train)

In [24]:
tfidf_test = tfidf_vectorizer.transform(x_test)

* __transform()__ : parameters generated from fit() method,applied upon model to generate transformed data set.

* __fit_transform()__ : combination of fit() and transform() api on same data set



### 7. Initializing PassiveAggressiveClassifier

In [21]:
from sklearn.linear_model import PassiveAggressiveClassifier
# Passive: if correct classification, keep the model; 
# Aggressive: if incorrect classification, update to adjust to this misclassified example.

In [25]:
pa_classifier = PassiveAggressiveClassifier(max_iter = 50)
pa_classifier.fit(tfidf_train, y_train)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

__max_iter__ --> Maximum number of iterations of the k-means algorithm for a single run.


### 8. Predicting and calculating the accuracy

In [33]:
from sklearn.metrics import accuracy_score

In [29]:
y_pred = pa_classifier.predict(tfidf_test)
score = accuracy_score(y_test, y_pred)

In [30]:
print('Accuracy:', score)

Accuracy: 0.9613782051282052


In [32]:
print(f'Accuracy: {round(score*100, 3)}%')

Accuracy: 96.138%


Now the accuracy of the model can be seen when it was conducting its tests. <br> Whereas we know the model's accuracy, we don't know the number of successful predictions/failures. <br> Therefore, we are now going to build a confusion matrix.

### 9. Building a confusion matrix

In [36]:
from sklearn.metrics import confusion_matrix

In [47]:
confusion_matrix(y_test, y_pred, labels = ['FAKE', 'REAL'])

array([[3001,  111],
       [ 130, 2998]], dtype=int64)

__Results:__ <br>
* The model successfully predicted 3001 positives. (Is fake and is predicted as fake)
* The model successfully predicted 2998 negatives. (Is real and is predicted as real)
* The model predicted 111 false positives. (Real news were considered as fake)
* The model predicted 130 false negatives. (Fake News were considered as real)

Additionally, we could compute the __F1-score__. <br> 
__F1 score__ is used to measure a test’s accuracy, therefore, an F-score is considered perfect when it's 1 , while the model is a total failure when it's 0.<br>  A good __F1-score__ means that you have low false positives and low false negatives,


F1 = 2TP / (2TP + FP + FN)

In [49]:
F1 = 2*3001 / (2*3001 + 130 + 111)
print(F1)

0.9613967643761012


## Part II

__The aim of part II is to experiment on the model by adding and substracting some variables and to see if the accuracy score is affected by it e.g.:__

1. Build a model that would take in only the title of a news article and then predict if it's fake or real.

2. Build a model that would take in both, the text of the article and the title

### 1. Experimenting with the title of news articles

In [74]:
labels_2 = df.label

In [75]:
x_train_title, x_test_title ,y_train_title ,y_test_title = train_test_split(df['title'].values.astype('str'), labels_2, test_size=0.3, random_state=7)


In [76]:
tfidf_vectorizer_title = TfidfVectorizer(stop_words='english', max_df=0.7)
#Fit and transform train set, transform test set

tfidf_train_title = tfidf_vectorizer_title.fit_transform(x_train_title) 
tfidf_test_title = tfidf_vectorizer_title.transform(x_test_title)

In [77]:
pa_classifier_title = PassiveAggressiveClassifier(max_iter = 50)
pa_classifier_title.fit(tfidf_train_title, y_train_title)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

In [79]:
y_pred_title = pa_classifier_title.predict(tfidf_test_title)
score_title = accuracy_score(y_test_title, y_pred_title)
print('Accuracy:', score_title)

Accuracy: 0.9299679487179487


In [80]:
print(f'Accuracy: {round(score_title*100, 3)}%')

Accuracy: 92.997%


In [82]:
confusion_matrix(y_test_title, y_pred_title, labels = ['FAKE', 'REAL'])

array([[2929,  183],
       [ 254, 2874]], dtype=int64)

__Results:__
* The model predicted 2929 positives (Is fake and were considered fake)
* The model predicted 2874 negatives (Is real and were considered )
* The model predicted 183 false positives ()
* The model predicted 254 false negatives ()